appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+		p.waitForSync(downstream, true)
+	}
+
+	return nil
+}
+
+// Backup returns a reader for streaming a backup in xbstream format.
+func (p *Process) Backup() (io.ReadCloser, error) {
+	cmd := exec.Command(
+		filepath.Join(p.BinDir, "innobackupex"),
+		"--defaults-file="+p.ConfigPath(),
+		"--host=127.0.0.1",
+		"--port="+p.Port,
+		"--user=flynn",
+		"--password="+p.Password,
+		"--socket=",
+		"--stream=xbstream",
+		".",
+	)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		stdout.Close()
+		return nil, err
+	}
+	return stdout, nil
+}
+
+// Restore restores the database from an xbstream backup.
+func (p *Process) Restore(r io.ReadCloser) error {
+	if err := p.writeConfig(configData{}); err != nil {
+		return err
+	}
+	if err := p.unpackXbstream(r); err != nil {
+		return err
+	}
+	if err := p.restoreApplyLog(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Process) unpackXbstream(r io.ReadCloser) error {
+	defer r.Close()
+
+	cmd := exec.Command(filepath.Join(p.BinDir, "xbstream"), "-x", "--directory="+p.DataDir)
+	cmd.Stdin = r
+
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("xbstream failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) restoreApplyLog() error {
+	cmd := exec.Command(
+		filepath.Join(p.BinDir, "innobackupex"),
+		"--defaults-file="+p.ConfigPath(),
+		"--apply-log",
+		p.DataDir,
+	)
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("innobackupex apply-log failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) assumeStandby(upstream, downstream *discoverd.Instance) error {
+	logger := p.Logger.New("fn", "assumeStandby", "upstream", upstream.Addr)
+	logger.Info("starting up as standby")
+
+	if err := p.writeConfig(configData{ReadOnly: true}); err != nil {
+		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+
+	if p.running() {
+		if err := p.stop(); err != nil {
+			return err
+		}
+	} else {
+		if p.WaitUpstream {
+			if err := p.waitForUpstream(upstream); err != nil {
+				return err
+			}
+		}
+
+		if err := func() error {
+			// TODO(benbjohnson): Lock table during backup: "FLUSH TABLES WITH READ LOCK"
+
+			logger.Info("retrieving backup")
+			resp, err := http.Get(fmt.Sprintf("http://%s/backup", upstream.Meta["HTTP_ADDR"]))
+			if err != nil {
+				logger.Error("error connecting to upstream for backup", "err", err)
+				return err
+			} else if resp.StatusCode != http.StatusOK {
+				resp.Body.Close()
+				logger.Error("error code returned from backup", "status_code", resp.StatusCode)
+				return err
+			}
+
+			logger.Info("restoring backup")
+			if err := p.Restore(resp.Body); err != nil {
+				logger.Error("error restoring backup", "err", err)
+				return err
+			}
+
+			// TODO(benbjohnson): Unlock tables after backup and log pos are retrieved.
```

Comment:
Because the lock is unecessary the unlock is too. However you raise a good point about retrieving the log position of the backup, we can actually do that without starting the server after we unpack the backup with `xbstream`. `xtrabackup` creates a file called `xtrabackup_binlog_info`, this will contain the binlog position.